Run dashboard as separate deployment

sleshchenko opened the pull request:
### What does this PR do?
It makes the Che operator run the dashboard as a separate Deployment. The same was previously done for helm charts.
The main reason why these changes were made, it's
> STEP1 - running the HTTP server that serves the dashboard in its own pod: customising the dashboard is just a matter of rebuilding the dashboard container image.
from https://github.com/eclipse/che/issues/15778#issuecomment-692535413

Now things changed and we have a plan to introduce a backend for the dashboard. So, it becomes a need.

It depends on https://github.com/eclipse-che/che-operator/pull/760


### Screenshot/screencast of this PR
<!-- Please include a screenshot or a screencast explaining what is doing this PR -->


### What issues does this PR fix or reference?
It's the first step for https://github.com/eclipse/che/issues/15778
It's the second step for https://github.com/eclipse/che/issues/17802
And a prerequisite for Dashboard backend (not sure if an issue exists at this point)


### How to test this PR?
Test that dashboard is run as dedicated deployment and is available on Che host.

I tested(and prepared a lot of fixes):

Minikube:
- multihost
- gateway singlehost
- native singlehost

OpenShift (RHPDS):
- multihost
- gateway singlehost

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse/che-operator/blob/master/README.md#updating-custom-resource-definition-file)
- [ ] [Nightly OLM bundle is up to date](https://github.com/eclipse/che-operator/blob/master/README.md#update-nightly-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

nickboldt commented:
> I made this draft to expose the made work but not sure when we should finalize and get it merged since it influences CRW productization, CRW should onboard +1 container.

Hard -10 from me. Why do we need Yet Another Tiny Container to have to productize, maintain, keep updated, stickhandle CVEs? What's the value for the customer here? Or for development/dogfooding? Or for dev sandbox users?

Today CRW is 25 images, wtih 2 more coming in 2.8. If we add async support in 2.9, that's 2 more. And then intellij support is another one. So that's 30 containers, and of those 30, **only 1/3rd are sidecars/lang support**.  That'll be _20 containers to JUST START CRW_, before even loading a workspace. 

Could we make this process simpler, not more complex? 

So... without a visible customer need or loud applause from @sympatheticmoose I don't immediately see the value in adding more technical debt for this.

sleshchenko commented:
@nickboldt I understand your pain but

>  Another Tiny Container

It's not going to be tiny in the future.
Dashboard Apache server will be extended soon with node backend, which would allow us to do Che + DevWorkspace Integration we want in a proper way.

Potentially, we can merge different processes (based on different technologies) into one container, but it's not the way K8s suggests us to go =)

Now we need to get on timeframe when CRW will pull the new image for dashboard, sorry about that.

nickboldt commented:
> merge different processes

Could we merge some of these?

* machine exec
* theia endpoint
* 2 brokers
* jwt proxy
* configbump
* traefik
* server
* operator

Maybe we can implement a "one in, one out" policy so that overall deployment complexity doesn't spiral out of control? Eg., you want dashboard to be a separate container, then configbump and traefik have to be combined.

sleshchenko commented:
> 2 brokers

we can merge 2 brokers, others are out of my team power, sorry )

sleshchenko commented:
Refactoring is moved to a separate PR https://github.com/eclipse-che/che-operator/pull/760

sleshchenko commented:
rebased against master and squashed fixups.

sleshchenko commented:
ci/prow/v7-devworkspace-happy-path fails with 

> time="2021-04-15T10:43:03Z" level=fatal msg="Failed to initialize default value: 'RELATED_IMAGE_dashboard'. Environment variable with default value was not found."

Seems it does not respect made changes from the PR https://github.com/eclipse-che/che-operator/pull/684/files#diff-1b647b08011e85965e46e70896e18383f07762e613ba677b5c7a9bf6ed8d221dR655

sleshchenko commented:
rebased against main branch and resolved conflicts - just regenerated nightly bundles with new version/dates.

tolusha commented:
/retest

sleshchenko commented:
Rebased against main branch. 
The only made change: fix to run on OpenShift https://github.com/eclipse-che/che-operator/compare/888544117b5dddad5c4415c90f5ce8dfd32f7e35..56e66d862e6dc2795b100ff194203352843723fa#diff-f26d61ce8d3fc7ba6d48229eb77344a2be0d209ac7b4eeb9f850e4d05873adacR129
```diff
-   if util.IsOpenShift {
+   if !util.IsOpenShift {
```

sleshchenko commented:
For some reason, events exposed by chectl server:logs are empty https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/eclipse-che_che-operator/684/pull-ci-eclipse-che-che-operator-main-v7-single-host-nightly-deployment/1384132729083269120/artifacts/single-host-nightly-deployment/single-host-nightly-deployment/artifacts/eclipse-che/events.txt
Which make impossible to figure out the tests failure reason.
For some unclear reasons it worked if I deploy with chectl server:deploy but did not work with opm (CatalogSource).

sleshchenko commented:
/retest

tolusha commented:
MInishift tests passed locally.

AndrienkoAleksandr commented on pkg/deploy/dashboard/dashboard.go:
```diff
@@ -0,0 +1,57 @@
+//
+// Copyright (c) 2012-2019 Red Hat, Inc.
```
Minor 2012-2019 -> 2021

sleshchenko replied:
I just copied pasted it from the existing files.
I would prefer not to touch it here not to retrigger all the tests again.

but in general it should be solved with automation, like devworkspace operator has https://github.com/devfile/devworkspace-operator/blob/main/Makefile#L242

sleshchenko replied:
fixed

AndrienkoAleksandr commented on pkg/deploy/dashboard/dashboard_deployment_test.go:
```diff
@@ -0,0 +1,114 @@
+//
+// Copyright (c) 2012-2019 Red Hat, Inc.
```
Minor 2021

tolusha commented:
/test v7-devworkspace-happy-path

sleshchenko commented:
Rebased against master to fix happy path tests.
Added fix into release scripts https://github.com/eclipse-che/che-operator/pull/684/files#diff-80b084421ff03b0d4874193917f225f185012ca25a072ea9aa94d38ef8a4ffefR88 Thank @tolusha for catching it.

tolusha commented:
/test v7-olm-nightly-deployment
